src/Compilers/CSharp/Test/Emit2/Semantics/LockTests.cs

```diff
@@ -0,0 +1,1102 @@
+        verifier.VerifyDiagnostics(
+            // (4,12): warning CS9214: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+            // ILock l1 = new Lock();
+            Diagnostic(ErrorCode.WRN_ConvertingLock, "new Lock()").WithLocation(4, 12),
+            // (7,12): warning CS9214: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+            // ILock l2 = new Lock();
+            Diagnostic(ErrorCode.WRN_ConvertingLock, "new Lock()").WithLocation(7, 12));
+    }
+
+    [Fact]
+    public void DerivedLock()
+    {
+        var source = """
+            using System;
+            using System.Threading;
+
+            DerivedLock l1 = new DerivedLock();
+            lock (l1) { Console.Write("1"); }
+
+            Lock l2 = l1;
+            lock (l2) { Console.Write("2"); }
+
+            DerivedLock l3 = (DerivedLock)l2;
+            lock (l3) { Console.Write("3"); }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public Scope EnterLockScope()
+                    {
+                        Console.Write("E");
+                        return new Scope();
+                    }
+
+                    public ref struct Scope
+                    {
+                        public void Dispose()
+                        {
+                            Console.Write("D");
+                        }
+                    }
+                }
+
+                public class DerivedLock : Lock { }
+            }
+            """;
+        var verifier = CompileAndVerify(source, verify: Verification.FailsILVerify,
+           expectedOutput: "1E2D3");
+        // Note: no warnings here as we don't expect `Lock` to be unsealed,
+        // so this doesn't warrant a warning in spec and implementation.
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void Downcast()
+    {
+        var source = """
+            using System;
+            using System.Threading;
+
+            object o = new Lock();
+            lock ((Lock)o) { Console.Write("L"); }
+            """;
+        var verifier = CompileAndVerify([source, LockTypeDefinition], verify: Verification.FailsILVerify,
+           expectedOutput: "ELD");
+        verifier.VerifyDiagnostics(
+            // 0.cs(4,12): warning CS9214: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+            // object o = new Lock();
+            Diagnostic(ErrorCode.WRN_ConvertingLock, "new Lock()").WithLocation(4, 12));
+    }
+
+    [Fact]
+    public void OtherConversions()
+    {
+        var source = """
+            #nullable enable
+            using System.Threading;
+
+            I i = new Lock();
```

Comment:
>I i = new Lock(); [](http://example.com/codeflow?start=12&length=17)

Are we testing scenario where `Lock` actually implements the interface?

Reply:
Yes, in test `CastToBase`.